Feat: 매거진 관리 : 삭제 기능

JJongsKim opened the pull request:
## 📮 관련 이슈
- Resolved: #이슈번호

## ⛳️ 작업 내용

매거진 페이지 삭제 기능을 위한 작업을 했습니다.
- [x] 매거진 다중 선택 UI 제작(이미 구현되어 있는 스크랩 다중 선택 참고)
- [x] 다중 선택 시 Link 영향을 받지 않도록 조건문을 통해 관리
- [x] 선택 버튼, 매거진 ids 배열을 recoil을 통해 관리 (선택을 취소할 시 ids 초기화, 선택했던 매거진 취소 시 ids에서 제외됨)
- [x] magazine delete API 호출 + 삭제 확인 완료

## 🌱 PR 포인트

## 📸 스크린샷
|스크린샷|
|:--:|
|![ezgif com-gif-maker](https://user-images.githubusercontent.com/81777778/211183560-d895f48f-426a-4cc9-8d89-6cb72bf0411b.gif)|

## 📎 레퍼런스

<!-- 참고한 레퍼런스가 있다면 기록해주세요 -->

JJongsKim committed:
:lipstick: Design: 매거진 멀티삭제 UI 스타일링

- 스크랩과 같은 형태로 멀티선택 스타일링
- 하위 tabMagazine으로 props 전달, 멀티선택 클릭 시 링크연결 막기
- 스크랩, 매거진 삭제버튼 클릭 시 바텀네비게이션 버그 발견 수정 필요

JJongsKim committed:
:sparkles: Feat: 매거진 삭제 기능 구현중

- test 완료 후 recoil로 수정하기
- selected 조건문 제대로 동작할 수 있도록 수정하기
- 위 두 사항 확인 후 delete api 테스트하기

JJongsKim committed:
:recycle: Refactor: recoil 선언, 다중선택 취소 시 빈 배열 전환

JJongsKim committed:
:sparkles: Feat: 다중선택 상태관리 확인완료

JJongsKim committed:
:wrench: Add: ids 배열 담기 selector로 수정

JJongsKim committed:
:sparkles: Feat: magazine delete API 호출 테스트코드 작성

- 호출 시 Ids number[] 값들이 넘어가지 않아 undefined로 들어가는 이슈 있음

JJongsKim committed:
:wrench: Add: mutation 파라미터 추가

- 선택된 매거진 값들이 삭제되었다가 안 되었다가 하는 이슈 발생

JJongsKim committed:
:wrench: Update: 매거진 코드 정리

- atom value 값이 제대로 넘어가지 않아 삭제가 되었다가 안 되었다가 반복되는 이슈가 있습니다.

JJongsKim committed:
Merge branch 'dev' into feat/magazine-delete

JJongsKim committed:
:bug: Fix: useEffect를 활용해 업데이트 된 atom, API로 전달

- atom 업데이트/전달 버그를 useEffect를 활용해 우선 수정하였습니다.
- useEffect에 추가한 eslint 무시 주석은 추후 수정이 필요합니다.

elbica commented on src/components/magazine/MagazineList/TabMagazine.tsx:
```diff
@@ -1,20 +1,39 @@
 }
 
-const TabMagazine = ({ magazines, selectItem, onScrollDown }: TabMagazineProps) => {
+const TabMagazine = ({ magazines, selectItem, option, onScrollDown }: TabMagazineProps) => {
   const ref = useScrollDetect<HTMLDivElement>({ onScroll: onScrollDown });
+  const boolOption = option;
```
option, boolOption이란 변수가 네이밍만 봐서 어떤 일을 하는지 알기 어렵네요🥲
`multiSelect` 같은 네이밍이 어떨지 제안합니다~

elbica commented on src/components/magazine/MagazineList/TabMagazine.tsx:
```diff
@@ -1,20 +1,39 @@
+  const boolOption = option;
+
+  // 썸네일 클릭 시 해당 id 추가, 선택 취소 시 id 확인 후 제거
+  const [magazineItems, setMagazineItems] = useRecoilState(magazineIdsArray);
+
+  const selectMagazineItems = useCallback(
+    (id: number) => {
+      magazineItems.includes(id)
+        ? setMagazineItems(magazineItems.filter((el) => el !== id))
+        : setMagazineItems((prev) => [...prev, id]);
+    },
+    // TODO 해당 경고 고민해보고 수정하기
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+    [magazineItems],
+  );
```
selector로 빼는 것도 좋아보여요
eslint 오류는 `[magazineItems, setMagazineItems]`로 디펜던시 넣으면 됩니다!

또한 `selectMagazineItems`는 인자로 들어온 id가 `magazineItems` 배열에 있으면 삭제하고, 반대로 없으면 추가하는 로직 같이 보이는데요

배열로 관리하는 것보다 `Set` 자료구조를 활용해 보시면 코드가 더 간단해질 것 같아요~

**예시 코드**
```tsx
// CategoryContentListMultiSelect.tsx
 const pickSet = useRef(new Set<number>());

pickSet.current.has(idx) ? pickSet.current.delete(idx) : pickSet.current.add(idx);

```

JJongsKim replied:
Set을 활용하는 방법도 있었군요 !! 👀

elbica commented on src/components/magazine/MagazineList/TabMagazine.tsx:
```diff
@@ -56,14 +75,26 @@ const TabMagazine = ({ magazines, selectItem, onScrollDown }: TabMagazineProps)
-                  height={'255px'}
-                />
-              </Link>
+              {boolOption ? (
+                <div>
+                  <Photo
+                    blur={<PhotoSelect enabled={selectItem} />}
+                    src={magazine.cover_url}
+                    width={'196px'}
+                    height={'255px'}
+                    onClick={() => selectMagazineItems(magazine.magazine_id)}
+                  />
+                </div>
+              ) : (
+                <Link href={`/magazine/${magazine.magazine_id}`}>
+                  <Photo
+                    blur={<PhotoSelect enabled={selectItem} />}
+                    src={magazine.cover_url}
+                    width={'196px'}
+                    height={'255px'}
+                  />
+                </Link>
+              )}
```
똑같은 `<Photo />` 컴포넌트가 반복되고 있네요!
```suggestion
            const Root = option ? <div /> : <Link href={`/magazine/${magazine.magazine_id}`} />;

          {cloneElement(Root, {
                  children: (
                    <Photo
                      blur={<PhotoSelect enabled={selectItem} />}
                      src={magazine.cover_url}
                      width={'196px'}
                      height={'255px'}
                      onClick={() => selectMagazineItems(magazine.magazine_id)}
                    />
                  ),
                })}
```
~이런식으로도 해결 가능해 보입니다!~
다시 보니 `onClick` props가 있었네요..!!🤣

JJongsKim replied:
cloneElement 사용해서 저렇게 줄일 수 있는건 오늘 처음 알았네요!!
앞으로도 잘 참고할게요! 
map 안에서 돌아가고 있는 magazine_id 때문에, Root 선언이 안 될 것 같아서 일단 원래코드 사용하겠습니당!😭

elbica commented on src/containers/magazine/TabMagazineViewContainer.tsx:
```diff
@@ -1,32 +1,106 @@
   const { magazines } = useGetMagazines();
+
+  const [selected, setSelected] = useState(initSelectedContext);
+  const [option, isOption] = useRecoilState(deleteOption);
```
여기도 `option` 같은 추상적인 네이밍보다 모르는 사람이 읽기만 해도 어떤일을 하는지 알 수 있게 구체적으로 네이밍 하는 것이 좋아보입니다!

elbica commented on src/containers/magazine/TabMagazineViewContainer.tsx:
```diff
@@ -1,32 +1,106 @@
+    if (deleteState === true) {
+      handleDeleteIds();
+    }
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [deleteState]);
```
`[deleteState, handleDeleteIds]`로 할 수 있을 것 같네요
eslint 오류 수정은 IDE 수준에서도 자동으로 고쳐주는 것으로 아는데 따로 eslint를 비활성시킨 이유가 있으실까요?

JJongsKim replied:
deleteState 이외에 eslint 오류를 통해 추가해야 하는 디펜던시도 넣고, 
코드를 수정하다보니 마지막엔 API 무한요청으로 넘어가버려서 우선은 eslint 비활성 주석을 달아 놓았습니다!

elbica replied:
지금 api 무한요청은 해결된 상태인데 eslint 비활성 주석은 예상치 못한 동작 방지를 위해서라도 없애는게 좋을 듯 합니다
혹시 아직도 무한요청이 발생하나요...?

JJongsKim replied:
제가 boolean state를 하나 놓치고 있었던 게 있어서 무한요청이 발생했었습니다!
해당 사항은 수정하였고, 디펜던시들도 추가하여 ESLint 오류없도록 만들어 다시 푸시했습니당😁

elbica commented on src/infra/api/magazine.ts:
```diff
@@ -19,6 +19,7 @@ class MagazineApi {
   };
   deleteMagazines = ({ ids }: DeleteMagazineRequest) => {
     //TODO flat ids array
+    console.log('삭제확인용', `/magazine?ids=${ids}`);
```
디버깅용 콘솔로그는 지우는게 좋아보입니다!

elbica reviewed (changes requested):
최근에 일정이 있어서 리뷰가 늦었네요..😥
최대한 다 이해해보고 코멘트 남겼습니다~

확인해 보시고 제가 남긴 것들 한 번 고민해 보시면 좋겠습니다
고생 많으셨습니다😊

JJongsKim committed:
:wrench: Update: 코드리뷰 수정사항 반영(변수 네이밍, 콘솔 삭제)

- 알아보기 힘든 변수 이름을 좀 더 예측하기 쉬운 이름으로 변경하였습니다. option/boolOption => selectDeleteOption/multiSelectOn
- magazine.ts deleteMagazines에 적어두었던  디버깅용 console.log를 삭제하였습니다.

JJongsKim committed:
:wrench: Update: 코드리뷰 수정사항 반영 (Set 자료구조 활용, 디펜던시 수정)

- 매거진 id 배열 관리에서 Set을 활용한 코드로 변경, 알맞은 디펜던시 추가

JJongsKim committed:
Merge branch 'dev' into feat/magazine-delete

elbica commented on src/components/magazine/MagazineList/TabMagazine.tsx:
```diff
@@ -1,20 +1,37 @@
+      pickSet.current.has(id) ? pickSet.current.delete(id) : pickSet.current.add(id);
+      setMagazineItems(Array.from(pickSet.current));
+    },
+    [pickSet, setMagazineItems],
```
리액트의 ref는 변경이 일어나도 디펜던시 배열로 추적이 안돼서 리렌더링이 일어나지 않습니다
state와 다르게 mutable한 데이터이기 때문입니다

set이 바껴도 렌더링 결과엔 반영 안될 수 있다는 의미인데, ref 자체를 props로 전달하지 않는 한 디펜던시 배열에 넣는건 의미가 없습니다! 아실지도 모르겠지만 추후 다른 기능 개발 하실때 관련 문제 겪으실까봐 말씀드려요

여기선 recoil 상태와 동기화 시켜주기만 하니까 기능 동작엔 문제는 없어 보이네요

elbica commented on src/containers/magazine/TabMagazineViewContainer.tsx:
```diff
@@ -1,32 +1,106 @@
 
 interface MagazineTabProps {
-  onMultiSelect?: () => void;
+  deleteOption?: boolean;
```
props 구조분해할당을 보니 실제로 쓰이지 않는 옵션같은데 삭제하는게 좋아보여요
Option 이라는 네이밍도 사실 제가 보기엔 유의미한 정보를 전달해 주지 않는 추상적인 네이밍 같아 보이긴 합니다

props 네이밍 하실때 일반적인 인터페이스는 유지해야 하지만 (value, onClick) 추상적인 단어(manager, process, xxxoption..)는 지양하는게 좋을 것 같아요!

JJongsKim replied:
아 어떻게 네이밍 할 때 효과적일지 관련해서도 공부를 좀 해야겠네요!
넵 반영해서 수정하겠습니당!

elbica commented on src/components/magazine/MagazineList/TabMagazine.tsx:
```diff
@@ -1,20 +1,37 @@
+  const multiSelectOn = selectDeleteOption;
+
+  // 썸네일 클릭 시 해당 id 추가, 선택 취소 시 id 확인 후 제거
+  const [, setMagazineItems] = useRecoilState(magazineIdsArray);
```
state 부분을 사용하고 있지 않으니useSetRecoilState를 쓰는게 좋겠네요!

elbica reviewed (commented):
코멘트 남겼으니 한번 더 확인 부탁드립니다~😊

JJongsKim committed:
:recycle: Refactor: 코드리뷰 수정사항 반

- 매거진, recoil 파일에서 추상적인 네이밍 변경
- 사용하지 않는 Props 제거
- useSetRecoilState() 변경

JJongsKim committed:
:bug: Fix: eslint 오류 수정 시 일어나는 무한요청버그 수정

- useEffect 디펜던시에 handleDeleteIds를 추가하여 ESLint 오류 해결
- useCallback을 활용하여 handleDeleteIds 함수를 감싸고 디펜던시들을 추가
- API 요청 작업이 끝난 후 deleteState의 state를 false로 변경하여 더 이상 요청되지 않도록 수정
